### PR TITLE
Add support for reverse PInvoke callouts to the JIT

### PR DESCRIPTION
**Pull request by jkotas**
Call reverse pinvoke enter/exit callouts in reverse pinvoke method prolog/epilog. This is fixing https://github.com/dotnet/corert/issues/611, no impact on CoreCLR.

### CODE REVIEW

**Comment by jkotas** [comment]
cc @pgavlin @dotnet/jit-contrib

File: src/jit/flowgraph.cpp [hunk 1]
```diff
@@ -8300,34 +8356,13 @@ void                Compiler::fgAddInternal()
...
-                                              tree,
-                                              oneReturnStmtNode);
-
-        }
-        else
-        {
-            oneReturnStmtNode = tree;
-        }
-
+        fgInsertStmtAtEnd(genReturnBB, tree);
```
Comment (jkotas) [inline src/jit/flowgraph.cpp]: This is a minor cleanup on x86-only path, to make handling of callout insertion uniform for different cases.

File: src/jit/flowgraph.cpp [hunk 2]
```diff
@@ -7808,6 +7807,65 @@ void                Compiler::fgConvertSyncReturnToLeave(BasicBlock* block)
...
+
+    // Add reverse pinvoke exit callout at the end of epilog
+
+    tree = gtNewOperNode(GT_ADDR, TYP_I_IMPL, gtNewLclvNode(lvaReversePInvokeFrameVar, TYP_BLK));
+
+    tree = gtNewHelperCallNode(CORINFO_HELP_JIT_REVERSE_PINVOKE_EXIT,
+        TYP_VOID, 0,
+        gtNewArgList(tree));
+
+    assert(genReturnBB);
```
Comment (pgavlin) [inline src/jit/flowgraph.cpp]: CC: `assert(genReturnBB != nullptr)`

File: src/jit/flowgraph.cpp [hunk 3]
```diff
@@ -7808,6 +7807,65 @@ void                Compiler::fgConvertSyncReturnToLeave(BasicBlock* block)
...
+*  Add enter/exit calls for reverse PInvoke methods
+*/
+
+void                Compiler::fgAddReversePInvokeEnterExit()
+{
+    assert(opts.IsReversePInvoke());
+
+    lvaReversePInvokeFrameVar = lvaGrabTempWithImplicitUse(false DEBUGARG("Reverse Pinvoke FrameVar"));
+
+    LclVarDsc * varDsc = &lvaTable[lvaReversePInvokeFrameVar];
```
Comment (pgavlin) [inline src/jit/flowgraph.cpp]: CC: `LclVarDsc* varDsc`

**Comment by pgavlin** [comment]
LGTM aside from a couple coding-convention-related nits.

File: src/jit/flowgraph.cpp [hunk 4]
```diff
@@ -7808,6 +7807,65 @@ void                Compiler::fgConvertSyncReturnToLeave(BasicBlock* block)
...
+
+    LclVarDsc* varDsc = &lvaTable[lvaReversePInvokeFrameVar];
+    varDsc->lvType = TYP_BLK;
+    varDsc->lvExactSize = eeGetEEInfo()->sizeOfReversePInvokeFrame;
+
+    GenTreePtr      tree;
+
+    // Add enter pinvoke exit callout at the start of prolog
+
+    tree = gtNewOperNode(GT_ADDR, TYP_I_IMPL, gtNewLclvNode(lvaReversePInvokeFrameVar, TYP_BLK));
```
Comment (sivarv) [inline src/jit/flowgraph.cpp]: (Minor) if you prefer, variable declaration and initialization could on the same line

GenTreePtr tree = gtNewOperNode(...);

File: src/jit/flowgraph.cpp [hunk 5]
```diff
@@ -7497,8 +7497,7 @@ GenTreePtr Compiler::fgGetCritSectOfStaticMethod()
 
 void                Compiler::fgAddSyncMethodEnterExit()
```
Comment (sivarv) [inline src/jit/flowgraph.cpp]: JIT CC: use standard function header

see https://github.com/dotnet/coreclr/blob/master/Documentation/coding-guidelines/clr-jit-coding-conventions.md#9.4

File: src/jit/flowgraph.cpp [hunk 6]
```diff
@@ -7808,6 +7807,65 @@ void                Compiler::fgConvertSyncReturnToLeave(BasicBlock* block)
 
 #endif // !_TARGET_X86_
 
+/*****************************************************************************
+*
+*  Add enter/exit calls for reverse PInvoke methods
+*/
+
+void                Compiler::fgAddReversePInvokeEnterExit()
```
Comment (sivarv) [inline src/jit/flowgraph.cpp]: JIT CC: please use standard function header. See
https://github.com/dotnet/coreclr/blob/master/Documentation/coding-guidelines/clr-jit-coding-conventions.md#9.4

**Comment by sivarv** [comment]
Looks good.

**Comment by briansull** [comment]
Looks good

**Pull request merged**